### PR TITLE
feat: `/play` subcommands (next, all, reverse, shuffle)

### PR DESCRIPTION
**Pull request by StaticRocket**
Setup basic play flags leveraging current PlayFlag enum system. Hopefully a good starting point for #51 

This adds the flags for:
- `PLAYALL` which overrides the URL logic to force the playlist to unravel for any URL that ytdl can interpret as a playlist
- `PLAYTOP` which replaces the current /playtop command

Please let me know if anything looks odd. This is my first real attempt at rust.

### CODE REVIEW

File: src/handlers/serenity.rs
Comment (aquelemiguel) [inline src/handlers/serenity.rs:144]: We could make use of [subcommands](https://discord.com/developers/docs/interactions/application-commands#example-walkthrough) here for the flags. Because right now this is a little finicky to interact with:

image

image

This grouping would be a little clearer and would allow for specific help text (e.g. **/play query top** with helper text **"Place track at the top of the queue"**):

image

**Review by aquelemiguel — CHANGES_REQUESTED** [review]
@StaticRocket Thank you for your submission. ☺️

Functionality-wise, it seems great. I just requested a few changes regarding subcommands, but they're open for discussion if anyone wants to chime in with an even better idea.

However, please be aware that #51 is already assigned to @rafaeldamasceno.

Before proceeding with this, let's understand whether these changes are compatible with the work Rafael might've already done. Let's respect the assignments. 🙂

File: src/handlers/serenity.rs
Comment (afonsojramos) [reply src/handlers/serenity.rs:144]: I don't know if I agree with you here @aquelemiguel, since this would pollute the command listing a bit. Additionally, the flag is optional, so you don't even need to type it.

Comment (joao-conde) [reply src/handlers/serenity.rs:144]: I like the suggested approach much more too.

Comment (aquelemiguel) [reply src/handlers/serenity.rs:144]: @afonsojramos A few considerations: 
* Frequently used commands always show up on top and there's autocomplete (plus we already had **/playtop** as a separate command and it was okay).
* UX-wise currently this is really awkward to interact with. Final command text looks weird and too long (if anything, it should just be `flag: top`) and there's no place for helper text.
* This is the expected use-case for subcommands.
 
I recommend pulling the PR and messing with it for a while.

Comment (StaticRocket) [reply src/handlers/serenity.rs:144]: Yeah, I agree with the remark about the helper text. That alone makes me want to change the implementation, but functionally I think I still prefer the flags.

Its quick and optional, but I wonder if there's some standard already in place to leverage instead of debating this single use case.

Comment (aquelemiguel) [reply src/handlers/serenity.rs:144]: > Yeah, I agree with the remark about the helper text. That alone makes me want to change the implementation, but functionally I think I still prefer the flags.

Yup, I think we can all agree flags are faster for old prefix-based commands.

> Its quick and optional, but I wonder if there's some standard already in place to leverage instead of debating this single use case.

For slash commands I'm pretty certain subcommands are the standard, please check [the docs](https://discord.com/developers/docs/interactions/application-commands#subcommands-and-subcommand-groups).
@afonsojramos Discord might nest subcommands in the future:

image

Comment (joao-conde) [reply src/handlers/serenity.rs:144]: And flags actually seem slower to use than a specific subcommand. Flags as a concept are useful in CLI applications, for slash commands I don't see it being coherent or useful. I much prefer the subcommand approach.

Comment (StaticRocket) [reply src/handlers/serenity.rs:144]: Alright, I'll start working on a subcommand implementation, but as unfamiliar with rust as I am, I am even more unfamiliar with serenity's subcommand system so I'll need another day or two. Are there any other things I should look into? Everyone happy with leveraging the PlayFlag enum?

Should we continue to let the interaction attributes override the command arguments or should we make that only work if the argument isn't explicitly given? Or should the argument that are being overridden be removed? I'm sure I'll have some input here when I go to implement it, but for future functionality is there anything that comes to mind?

**Comment by StaticRocket** [comment]
Well this all works now, but there's an issue with the way the playlist spooling system works right now. Queuing an instant mix will result in an infinite queue, adding results until the bot is restarted. This continues even if the bot is forced to leave, or at least the yt-dlp thread continues after the bot is forced to leave. The queue seems to stop growing after the `/leave` command is issued.

**Comment by joao-conde** [comment]
@StaticRocket I am unfamiliar with what an instant mix is. If you can fill me in I can propose some solution. 

I know that for live streams we check the track metadata for the duration of the track.

In any case, I will review once there are no known bugs/issues with the PR.

**Comment by aquelemiguel** [comment]
@StaticRocket I'm also unfamiliar with what instant mixes are, but I'm assuming it's like those infinitely generating playlists Spotify has? Could you please provide an example link?

**Comment by StaticRocket** [comment]
Sure, here's a link to a YouTube instant mix: https://youtu.be/1V_xRb0x9aw?list=RD1V_xRb0x9aw

And what they look like:
image

The only reason we haven't had to deal with this issue before now is because these don't quite work like a playlist. They cannot be accessed by the usual playlist formatted URL.

https://www.youtube.com/playlist?list=RD1V_xRb0x9aw

But you'll find that when you reach the last song in the playlist the size will increase as it adds more videos.

This is probably something we should use those new ytdl parameters for, but I'm not sure what we should do. Having a hard cutoff for playlist lengths seems bad. Having something like a rate-limit that only applies to instant mixes seems alright, but we still have to deal with the runaway process, assuming that wasn't just my machine being weird.

I assume the ideal thing would be to wait until the current video is half way done before processing the next video in the mix (making this a new enqueuing method just for mixes and infinite length playlists), but I haven't looked to see how feasible that is yet.

**Comment by joao-conde** [comment]
So a mix is kind of youtube suggestions, in a playlist, but when you reach the end it just gives you new ones? Ok, if that is so, I would say we either:

a) don't allow queueing mixes
b) queue a frozen snapshot of the mix, meaning we take the first tracks and queue them, ignoring newly added ones

I will take a look at the source later to suggest such an implementation but I believe it would be possible.

**Comment by aquelemiguel** [comment]
@StaticRocket I just checked and yt-dlp seems to return 256 songs with an instant mix. I wouldn't worry much about this, TBH.

Alternatively, the simplest fix would be limiting the number of songs, either by trimming yt-dlp's output on the play command or by passing a custom yt-dlp argument (possibly **--playlist-end**). But again, I personally wouldn't bother, I don't know what the rest think.

**Comment by StaticRocket** [comment]
I'm comfortable with leaving it, but currently the /stop and /clear commands can't do anything to stop the queueing until all 256 entries have been added.

Not sure if people will like that.

**Review by aquelemiguel — COMMENTED** [review]
This is working pretty well! ✨
Before merging, don't forget to run `cargo fmt` so our pipeline passes.

**Comment by aquelemiguel** [comment]
> Currently the /stop and /clear commands can't do anything to stop the queueing until all 256 entries have been added.

@StaticRocket This is true (for any playlist really) and not ideal. Because AFAIK communication between commands can only be accomplished through `context.data`, I can only really see a solution that would involve reading some sort of "stop flag". Set it to true in `/stop` and have `/play` break the queue loop if it's set.

I don't think this is super relevant to this issue in particular but definitely something to think about in the future.

File: src/handlers/serenity.rs
Comment (aquelemiguel) [inline src/handlers/serenity.rs:133]: Sorry, just a final change, I missed this. 😓

Please rename default to **end** and change description to **"Add a track to the end of the queue"** - it'll make it more in line with the rest of the descriptions.

**Review by aquelemiguel — CHANGES_REQUESTED** [review]

**Review by aquelemiguel — APPROVED** [review]
@StaticRocket LGTM, thank you for your work and insights. 🥰

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:24]: Now that we are it, let's change the `_play` signature to take `flag: PlayFlag` instead of a reference (does not really make sense to pass a reference for an enum value)

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:49]: make the default case return the `PlayFlag::END` maybe (and for clarity add a "end" => PlayFlag::END" too)

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:71]: Can be rewritten as:

```rust
_  if url.contains("youtube.com/playlist?list=") => EnqueueType::PLAYLIST,
_  if url.starts_with("http") => EnqueueType::URI,
_ => EnqueueType::SEARCH
```

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:97]: This is the PlayFlag::END right? Make it explicit.

File: src/commands/mod.rs
Comment (joao-conde) [inline src/commands/mod.rs:22]: While we are at it, refactor this enum to be something like `PlayMode` what do you think @aquelemiguel ? Also, usually enum values are PascalCase (so All, End, Next).

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:156]: Change `enqueue_song` too to take a owned value of the play flag instead of a reference.

**Review by joao-conde — CHANGES_REQUESTED** [review]
@aquelemiguel tell me what you think of these changes if you agree to.

@StaticRocket let me know if anything is unclear. Don't forget to run `cargo fmt` and `cargo clippy`

File: src/commands/mod.rs
Comment (aquelemiguel) [reply src/commands/mod.rs:22]: Yeah, I like it, let's drop the flag naming.

**Comment by aquelemiguel** [comment]
@joao-conde Agree with all your takes! 🚀

File: src/commands/play.rs
Comment (StaticRocket) [reply src/commands/play.rs:49]: Yeah, I should do that. For some reason I thought I should allow for a case where the passed flag takes priority, in case some command decides to wrap the play command. In hindsight that functionality should be removed as the play command is now very dependent on message context.

Comment (StaticRocket) [reply src/commands/play.rs:49]: @joao-conde something I forgot about this is that it currently covers the '&_' case. Modifying this to specifically include the "end" case will make the pattern non-exhaustive.

File: src/commands/mod.rs
Comment (afonsojramos) [reply src/commands/mod.rs:22]: Circling back to this conversation due to the PascalCase enums. In Rust, Enums are in Pascal case, which we can already find in our code in the `Connection` enum. However, it is true that the old `EnqueueType` and `PlayFlag` were all capitalised... Either way, we should move to the standard imo.

**Comment by aquelemiguel** [comment]
@StaticRocket You're aiming to implement the rest of the subcommands on this PR? I
If so, I'll leave it open.

File: src/handlers/serenity.rs
Comment (aquelemiguel) [inline src/handlers/serenity.rs:172]: Not quite, should work for both video/playlist links and simple playlist links.
For simplicity's sake, reword this to **"Add a playlist to the queue in reverse order"**.

File: src/handlers/serenity.rs
Comment (aquelemiguel) [inline src/handlers/serenity.rs:185]: And here **"Add a playlist to the queue in random order"**.

File: src/commands/play.rs
Comment (aquelemiguel) [inline src/commands/play.rs:207]: Good catch! 🚀

**Review by aquelemiguel — CHANGES_REQUESTED** [review]

**Review by afonsojramos — CHANGES_REQUESTED** [review]

File: src/commands/play.rs
Comment (afonsojramos) [reply src/commands/play.rs:152]: I like this implementation actually, it looks very clean for what it is. What are the benefits of one over the other @aquelemiguel?

File: src/commands/play.rs
Comment (afonsojramos) [reply src/commands/play.rs:207]: Indeed! But maybe we could add some text to explain what happened?

File: src/commands/play.rs
Comment (aquelemiguel) [reply src/commands/play.rs:152]: Yeah, I suggested that in my review above, it's cleaner to let yt-dlp do the playlist manipulation work.

**Comment by rafaeldamasceno** [comment]
@StaticRocket Thank you for your great work on these commands!

As for the feature and as a discussion topic for everyone, I have some thoughts I want to add, even though I have probably missed this train. When I first suggested the flags feature, I had in mind the possibility of mixing and matching flags (for example, shuffling a playlist and playing it next with -sn, playing all the videos in a video with playlist instantly with -aj, etc.). 

Is there any plan for mixing and matching flags that doesn't result in a subcommand explosion?

File: src/commands/play.rs
Comment (aquelemiguel) [reply src/commands/play.rs:207]: @afonsojramos You mean a Discord message to inform the user?
Something like **"An error occurred while trying to queue track `XXXXX`! This video might be privated."**?

File: src/commands/play.rs
Comment (aquelemiguel) [reply src/commands/play.rs:152]: @afonsojramos This block is just reordering the playlist but we ask yt-dlp to return it already correctly ordered.
So we can delete this code and pass this transformation overhead to yt-dlp (which is better, even if negligible).

**Comment by aquelemiguel** [comment]
> Is there any plan for mixing and matching flags that doesn't result in a subcommand explosion?

I thought about that too, but I believe right now this isn't possible with slash commands. The optional flags approach was finicky even with only one flag and with subcommands you'd have to cover all possible flags combinations.

I think here we should try to cover the most popular cases and think of this less like flags and more like subcommands. It's better to have a few stable, intuitive subcommands instead of a dodgy fully flexible flag implementation is what I mean. It does suck to lose some functionality, but following Discord standards benefits us in the long run.

**Comment by afonsojramos** [comment]
> I think here we should try to cover the most popular cases and think of this less like flags and more like subcommands. It's better to have a few stable, intuitive subcommands instead of a dodgy fully flexible flag implementation is what I mean. It does suck to lose some functionality, but following Discord standards benefits us in the long run.

Agree with @aquelemiguel here.

**Comment by rafaeldamasceno** [comment]
> It's better to have a few stable, intuitive subcommands instead of a dodgy fully flexible flag implementation is what I mean.

While I undestand what you mean, I don't believe the latter option would be dodgy or unintuitive. If manually parsing a flags string parameter is undesirable (and I can agree with that), having the flags as optional booleans in the play command would still allowed users to know what they were used for and for them to activate only the ones they wanted to use. 

The main disadvantage I see with the subcommands approach is that it doesn't lose *some* functionality, but instead most of the flexibility that the flags system would provide. Currenly, I can do all of the behavior of these commands by manually doing the work after queueing something, and since they only provide one functionality, it's not that much work (i.e. I'm not sure how much additional value these subcommands provide, since they only remove the need of calling a single command afterwards). The purpose of the flags would be to have any of the desired behaviors to be done automatically when enqueueing someting. I understand that more subcommands can be added in the future (some for shuffling and playing the playlist next, or any other combinations that are deemed common or useful), but IMHO it's only a matter of time before a combination that wasn't implemented is requested because it's used a lot by a specific group or person.

**Comment by aquelemiguel** [comment]
> If manually parsing a flags string parameter is undesirable

@rafaeldamasceno Something like `/play query flags -a -r` where flags is a string option?
It mixes application commands and CLI interaction but I don't think it's atrocious. 🤷 

> having the flags as optional booleans in the play command

This would look terrible, no? For example, `/play query all true random true`. Options must always have a value.

> Currently, I can do all of the behavior of these commands by manually doing the work after queueing something, and since they only provide one functionality, it's not that much work

Sort of, play with the random flag is similar to playing and then shuffling, but if you had other tracks in the queue, they'd also be mixed with the playlist. But I see what you mean, they're not fantastic shortcuts.

> I understand that more subcommands can be added in the future (some for shuffling and playing the playlist next, or any other combinations that are deemed common or useful), but IMHO it's only a matter of time before a combination that wasn't implemented is requested because it's used a lot by a specific group or person.

I agree, but it's a little difficult for me to picture how useful these combinations are because I never really used flags on other bots. 
Let's ensure here on #140 that all the flags work well separately via subcommands. Then I suggest opening another issue for discussing how we'd go about mixing and matching flags in an intuitive way.

**Comment by rafaeldamasceno** [comment]
@afonsojramos For sure, it's just that this PR isn't really aligned with what I was expecting in #51 and the discussion then started here. I will continue the discussion there.

In any case, @aquelemiguel the next steps you proposed sounds great. What I then recommend is that #51 isn't closed by this PR (although very advanced by it) and that this development takes into account ways to make it easy to invoke internal functions with different flags (e.g `enqueue_song` was changed to receive a single `PlayMode` and matches only on one `PlayMode` when there could be multiple, `PlayMode` could indeed just be called `PlayFlag` again, etc.). Otherwise the current rework of the playing/enqueing will need to be undone almost straight away when implementing multiple flags.

File: src/commands/play.rs
Comment (joao-conde) [reply src/commands/play.rs]: Ye, use both, that's what I meant, just for clarity, like

```rust
"end" => PlayMode::End,
_ => PlayMode::End
```

File: src/commands/play.rs
Comment (joao-conde) [reply src/commands/play.rs:207]: The println right now does not make much sense, as this prints to the server side and we don't have server logs (which in the future could be something we look into, using a logger).

Sending a message for each private track to the discord server might be too much. I would say just fail silently (return only).

Comment (afonsojramos) [reply src/commands/play.rs:208]: I actually would leave the print so we can easily replace with the logger in the future.

**Review by afonsojramos — APPROVED** [review]
Approving since the PR is already in a mergeable state. We can update the Shuffle and Reverse changes in another PR or in this one, up to you @StaticRocket, in my opinion.

**Comment by aquelemiguel** [comment]
> Approving since the PR is already in a mergeable state. We can update the Shuffle and Reverse changes in another PR or in this one, up to you @StaticRocket, in my opinion.

Sure, that and the jump subcommand is missing but can also be tackled in a different one.

**Comment by StaticRocket** [comment]
> the jump subcommand is missing

Absolutely going to leave that to someone else. It's not a bad feature, but I don't see myself using it that much. Implementation wise it should be easy; just make a call to `/stop` before queuing anything.

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:68]: `enqueue_song`/`enqueue_playlist` should take an owned value for the mode.

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:76]: Make `calculate_time_until_play` take ownership of the mode. You will get "moved value" errors but you can derive the Copy trait for these enums since they are so "light".

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:80]: Not sure, but try matching on a tuple, like:

```rust
match (enqueue_type, mode) {
  (EnqueueType::Link | EnqueueType::Search, PlayMode::Next) => ...
}
```

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:109]: Take ownership of mode.

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:143]: Take ownership of mode.

File: src/commands/play.rs
Comment (joao-conde) [inline src/commands/play.rs:183]: Idem

**Review by joao-conde — CHANGES_REQUESTED** [review]
LGTM, make the small changes requested and I approve it.

**Review by aquelemiguel — APPROVED** [review]
Reviewed with @joao-conde, LGTM as well. ☺️

**Review by afonsojramos — APPROVED** [review]

**Comment by afonsojramos** [comment]
Awesome PR, and awesome contributions @StaticRocket!! 🚀

**Review by joao-conde — APPROVED** [review]
LGTM

**Pull request merged**
